File: src/console/command/parameter/def/StringEnumParameter.php
<?php

namespace pocketcloud\cloud\console\command\parameter\def;

use pocketcloud\cloud\console\command\parameter\CommandParameter;
use pocketcloud\cloud\console\command\parameter\exception\ArgumentParseException;

readonly class StringEnumParameter extends CommandParameter {

    protected array $allowedStrings;

    public function __construct(
        string $name,
        array $allowedStrings,
        protected bool $caseSensitive,
        bool $optional,
        protected ?string $typeName = null,
        ?string $customErrorMessage = null
    ) {
        parent::__construct($name, $optional, $customErrorMessage);
        $this->allowedStrings = array_map(function (string $string): string {
            return $this->caseSensitive ? $string : strtolower($string);
        }, $allowedStrings);
    }

    public function parseValue(string $input): string {
        $input = $this->caseSensitive ? $input : strtolower($input);
        if (!$this->isAllowedString($input)) throw new ArgumentParseException("Given string is not allowed inside this StringEnumArgument");
        return $input;
    }

    public function onTabCompleteMatch(string $currentArg): array {
        if ($currentArg == "") return $this->allowedStrings;
        return array_values(array_filter(
            $this->allowedStrings,
            fn (string $string) => str_contains($string, $currentArg)
        ));
    }

    public function getType(): string {
        if ($this->typeName !== null) return $this->typeName;
        $allowedStrings = $this->allowedStrings;
        if (count($allowedStrings) > 2) {
            return implode("|", array_slice($allowedStrings, 0, 2)) . "|...";
        }

        return implode("|", $this->allowedStrings);
    }

    public function getAllowedStrings(): array {
        return $this->allowedStrings;
    }

    public function isAllowedString(string $string): bool {
        return in_array($this->caseSensitive ? $string : strtolower($string), $this->allowedStrings);
    }

    public function isCaseSensitive(): bool {
        return $this->caseSensitive;
    }
}

File: cloud-core/src/main/java/de/pocketcloud/cloud/console/command/parameter/def/StringEnumParameter.java
package de.pocketcloud.cloud.console.command.parameter.def;

import de.pocketcloud.cloud.console.command.parameter.BaseCommandParameter;
import de.pocketcloud.cloud.console.command.exception.ArgumentParseException;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public class StringEnumParameter extends BaseCommandParameter {

    private final List<String> allowedStrings;
    private final boolean caseSensitive;
    private final String typeName;

    public StringEnumParameter(String name, List<String> allowedStrings, boolean caseSensitive, boolean optional, String typeName, String customErrorMessage) {
        super(name, optional, customErrorMessage);
        this.caseSensitive = caseSensitive;
        this.typeName = typeName;
        this.allowedStrings = allowedStrings.stream()
                .map(s -> caseSensitive ? s : s.toLowerCase())
                .collect(Collectors.toList());
    }

    public StringEnumParameter(String name, List<String> allowedStrings, boolean caseSensitive, boolean optional) {
        this(name, allowedStrings, caseSensitive, optional, null, null);
    }

    public StringEnumParameter(String name, boolean optional, String... allowedStrings) {
        this(name, Arrays.asList(allowedStrings), false, optional, null, null);
    }

    @Override
    public String parseValue(String input) throws ArgumentParseException {
        String normalised = caseSensitive ? input : input.toLowerCase();
        if (!allowedStrings.contains(normalised)) throw new ArgumentParseException("Given string is not allowed");
        return normalised;
    }

    @Override
    public List<String> onTabCompleteMatch(String currentArg) {
        if (currentArg.isEmpty()) return List.copyOf(allowedStrings);
        String needle = caseSensitive ? currentArg : currentArg.toLowerCase();
        return allowedStrings.stream().filter(s -> s.contains(needle)).collect(Collectors.toList());
    }

    @Override
    public String getType() {
        if (typeName != null) return typeName;
        if (allowedStrings.size() > 2)
            return allowedStrings.get(0) + "|" + allowedStrings.get(1) + "|...";
        return String.join("|", allowedStrings);
    }

    public List<String> getAllowedStrings() {
        return List.copyOf(allowedStrings);
    }

    public boolean isAllowedString(String s) {
        return allowedStrings.contains(caseSensitive ? s : s.toLowerCase());
    }
}